src/servers/src/otlp.rs

```diff
@@ -0,0 +1,343 @@
+/// See
+/// https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L162
+/// for data structure of OTLP metrics.
+///
+/// Returns `InsertRequests` and total number of rows to ingest
+pub fn to_grpc_insert_requests(
+    request: ExportMetricsServiceRequest,
+) -> Result<(InsertRequests, usize)> {
+    let metrics = request
+        .resource_metrics
+        .iter()
+        .flat_map(|resource_metrics| &resource_metrics.scope_metrics)
+        .flat_map(|scope_metrics| &scope_metrics.metrics);
+
+    let mut insert_batch = Vec::with_capacity(metrics.size_hint().0);
+    for metric in metrics {
+        if let Some(insert) = encode_metrics(metric)? {
+            insert_batch.push(insert);
+        }
+    }
+
+    let rows = insert_batch.iter().map(|i| i.row_count as usize).sum();
+    let inserts = InsertRequests {
+        inserts: insert_batch,
+    };
+
+    Ok((inserts, rows))
+}
+
+fn encode_metrics(metric: &Metric) -> Result<Option<InsertRequest>> {
+    let name = &metric.name;
+    // note that we don't store description or unit, we might want to deal with
+    // these fields in the future.
+    if let Some(data) = &metric.data {
+        match data {
+            metric::Data::Gauge(gauge) => encode_gauge(name, gauge).map(Some),
+            metric::Data::Sum(sum) => encode_sum(name, sum).map(Some),
+            metric::Data::Summary(summary) => encode_summary(name, summary).map(Some),
+            // TODO(sunng87) leave histogram for next release
+            metric::Data::Histogram(_hist) => Ok(None),
+            metric::Data::ExponentialHistogram(_hist) => Ok(None),
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+fn write_attribute(lines: &mut LinesWriter, attr: &KeyValue) -> Result<()> {
+    if let Some(val) = attr.value.as_ref().and_then(|v| v.value.as_ref()) {
+        match val {
+            any_value::Value::StringValue(s) => lines
+                .write_tag(&attr.key, s)
+                .context(error::OtlpMetricsWriteSnafu)?,
+
+            any_value::Value::IntValue(v) => lines
+                .write_tag(&normalize_otlp_name(&attr.key), &v.to_string())
+                .context(error::OtlpMetricsWriteSnafu)?,
+            any_value::Value::DoubleValue(v) => lines
+                .write_tag(&normalize_otlp_name(&attr.key), &v.to_string())
+                .context(error::OtlpMetricsWriteSnafu)?,
+            // TODO(sunng87): allow different type of values
+            _ => {}
+        }
+    }
+
+    Ok(())
+}
+
+fn write_timestamp(lines: &mut LinesWriter, time_nano: i64) -> Result<()> {
+    lines
+        .write_ts(GREPTIME_TIMESTAMP, (time_nano, Precision::Nanosecond))
+        .context(error::OtlpMetricsWriteSnafu)?;
+    Ok(())
+}
+
+/// encode this gauge metric
+///
+/// note that there can be multiple data points in the request, it's going to be
+/// stored as multiple rows
+fn encode_gauge(name: &str, gauge: &Gauge) -> Result<InsertRequest> {
```

Comment:
I found `encode_gauge` and `encode_sum` have the common code that could be extracted.
Unless that was done deliberately.